### DISCUSSION TITLE
Unsupported MS SQL localdb via pyodbc

### DISCUSSION THREAD

**Discussion by epizut** [sqlalchemy/sqlalchemy#8594] [Usage Questions] [answered]
### Describe the use case

Microsoft SQL Server localdb connect string is unusable with the current pyodbc connector.
MSSQL localdb uses a local named pipe connection were you must not set any database name.
Unfortunately the current pyodbc connector force using _DSN_ instead of _SERVER_ when the database name is empty.

Here is the problematic code: [connectors/pyodbc.py](https://github.com/sqlalchemy/sqlalchemy/blob/befa8e92675dea992423011b929dfc7ca45de6b6/lib/sqlalchemy/connectors/pyodbc.py#L89)

### Databases / Backends / Drivers targeted

mssql+pyodbc

### Example Use

Unfortunatly
`create_engine('mssql+pyodbc://(localdb)\MSSQLLocalDB/?driver=ODBC+Driver+17+for+SQL+Server')`
is geratating:
`'dsn=(localdb)\\MSSQLLocalDB;Trusted_Connection=Yes;driver=ODBC Driver 17 for SQL Server'`
instead of:
`'DRIVER={ODBC Driver 17 for SQL Server};Server=(localdb)\MSSQLLocalDB;Integrated Security=true'`

**Comment by zzzeek** [comment, marked as answer]
hi -

use [pass through exact pyodbc string](https://docs.sqlalchemy.org/en/14/dialects/mssql.html#pass-through-exact-pyodbc-string) for this

```py
>>> from sqlalchemy.engine import URL
>>> connection_string = "DRIVER={ODBC Driver 17 for SQL Server};Server=(localdb)\MSSQLLocalDB;Integrated Security=true'"
>>> connection_url = URL.create("mssql+pyodbc", query={"odbc_connect": connection_string})
>>> from sqlalchemy.dialects.mssql import pyodbc
>>> print(pyodbc.dialect().create_connect_args(connection_url))
[["DRIVER={ODBC Driver 17 for SQL Server};Server=(localdb)\\MSSQLLocalDB;Integrated Security=true'"], {}]
```

for an all in one URL string, we can encode the above:

```py
>>> import urllib.parse
>>> urllib.parse.quote_plus(connection_string)
'DRIVER%3D%7BODBC+Driver+17+for+SQL+Server%7D%3BServer%3D%28localdb%29%5CMSSQLLocalDB%3BIntegrated+Security%3Dtrue%27'
>>> url = f"mssql+pyodbc://?odbc_connect={urllib.parse.quote_plus(connection_string)}"
>>> url
'mssql+pyodbc://?odbc_connect=DRIVER%3D%7BODBC+Driver+17+for+SQL+Server%7D%3BServer%3D%28localdb%29%5CMSSQLLocalDB%3BIntegrated+Security%3Dtrue%27'
>>> from sqlalchemy.engine import make_url
>>> print(pyodbc.dialect().create_connect_args(make_url(url)))
[["DRIVER={ODBC Driver 17 for SQL Server};Server=(localdb)\\MSSQLLocalDB;Integrated Security=true'"], {}]

```

**Reply by epizut** [reply]
thanks a lot!
